compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -7164,15 +7194,39 @@ private void createTypeCastExprForRelationalExpr(BLangBinaryExpr binaryExpr, int
         }
 
         if (isLhsStringType && !isRhsStringType) {
-            binaryExpr.rhsExpr = createTypeCastExpr(binaryExpr.rhsExpr, symTable.stringType);
+            addTypeCastForBinaryExprB(binaryExpr, symTable.stringType, rhsExprType);
             return;
         }
 
         if (!isLhsStringType && isRhsStringType) {
-            binaryExpr.lhsExpr = createTypeCastExpr(binaryExpr.lhsExpr, symTable.stringType);
+            addTypeCastForBinaryExprA(binaryExpr, symTable.stringType, lhsExprType);
         }
     }
 
+    private void addTypeCastForBinaryExprA(BLangBinaryExpr binaryExpr, BType rhsExprType, BType lhsExprType) {
+        if (lhsExprType.tag == TypeTags.UNION && lhsExprType.isNullable()) {
+            binaryExpr.rhsExpr = addNilType(rhsExprType, binaryExpr.rhsExpr);
+        } else {
+            binaryExpr.lhsExpr = createTypeCastExpr(binaryExpr.lhsExpr, rhsExprType);
+        }
+    }
+
+    private void addTypeCastForBinaryExprB(BLangBinaryExpr binaryExpr, BType lhsExprType, BType rhsExprType) {
+        if (rhsExprType.tag == TypeTags.UNION && rhsExprType.isNullable()) {
+            binaryExpr.lhsExpr = addNilType(lhsExprType, binaryExpr.lhsExpr);
```

Comment:
Can you please explain what we are trying to do here? Aren't we creating new types even if already nilable?

Reply:
> Can you please explain what we are trying to do here?

Basically, I am trying to make similar both `LHS` and `RHS` types in binary expression. Otherwise, there will be an `ArrayIndexOutOfBoundsException` at codegen phase. 
In this particular `if block` will check whether the type of the `rhsExpr` is nilable and if so it will make the type of the `lhsExpr` also nilable.

> Aren't we creating new types even if already nilable?

No. We won't create new types for already nilable expressions.